Enable User customization of OpenAPI UI Endpoint

abutch3r opened the pull request:
Allow user to define a property for the MP OpenAPI UI Endpoint such that the UI is available on that path e.g. `/custom/location` 

Given package is used by all OpenAPI feature versions, it will always default to `/openapi/ui` and is currently beta guarded

includes validation to prevent conflict with the OpenAPI 

Expands UI testing to include test for updating of UI - runs the same tests against the UI as the basic test and basic test covers the reverting to default 

#25072

abutch3r committed:
Support OpenAPI UI context path customization

Simplify class and move WAB Manager class into its own file.
Add unit tests to cover path maniuplation and a small number of validation scenarios.
Improve validation by checking path can construct a valid URL.
if path fails for multiple reasons - check them all instead of outright failure

Improve messages and add some new messages

Add Beta and custom system property guards and update test to cover these requirements.

Update Fat test to use own server definition to make it easier for future tests to update the endpoint value and other updates,

Enable customization of OpenAPI UI Endpoint

Add test case that checks when moved that the UI still renders correctly

abutch3r commented:
#libby 
#build

ramkumar-k-9286 commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+
+# Message prefix block: CWWKO1750 - CWWKO1799
+
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED=CWWKO1750W: The OpenAPI UI Path cannot be set to {0}. This Path is reserved for use by the application server.
```
```suggestion
CWWKO1750W_OPEN_API_UI_PATH_RESERVED=CWWKO1750W: The application server reserves the OpenAPI UI Path, so it cannot be set to {0}.
```

helyarp replied:
I prefer the original message, with lowercase p unless path needs to be Path:
The OpenAPI UI Path cannot be set to {0}. This path is reserved for use by the application server.

abutch3r replied:
This message is used solely for one other endpoint - so have focused the message on that conflict instead of a generic message for any potential endpoint value

ramkumar-k-9286 commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+# Message prefix block: CWWKO1750 - CWWKO1799
+
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED=CWWKO1750W: The OpenAPI UI Path cannot be set to {0}. This Path is reserved for use by the application server.
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.explanation=The OpenAPI UI Path cannot be set to a value that has been reserved for use by the application server.
```
```suggestion
CWWKO1750W_OPEN_API_UI_PATH_RESERVED.explanation= The application server reserves the use of certain values, preventing the OpenAPI UI Path from being set to these values.
```

helyarp replied:
Although we try not to have passive voice in messages, I prefer the original message with "has been reserved" changed to "is reserved":

The OpenAPI UI Path cannot be set to a value that is reserved for use by the application server.

abutch3r replied:
This message is used solely for one other endpoint - so have focused the message on that conflict instead of a generic message for any potential endpoint value

ramkumar-k-9286 commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.useraction=Specify a valid value for the extensions attribute in the Micro Profile configuration.
+
+CWWKO1751W_OPEN_API_UI_PATH_INVALID=CWWKO1751W: The OpenAPI UI Path cannot be set to {0}. The Path is invalid or contains invalid characters.
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.explanation=The OpenAPI UI Path cannot be set to a value containing invalid characters.
```
```suggestion
CWWKO1751W_OPEN_API_UI_PATH_INVALID.explanation= The value assigned to the OpenAPI UI Path must not include any characters that are considered invalid.

helyarp replied:
Suggested message explanation:
The OpenAPI UI Path cannot be set to a value that contains invalid characters.

abutch3r replied:
Updated as suggested

ramkumar-k-9286 commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.explanation=The OpenAPI UI Path cannot be set to a value containing invalid characters.
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.useraction=Specify a valid value for the extensions attribute in the configuration.
+
+CWWKO1752W_OPEN_API_UI_PATH_POTENTIAL_CONFLICT=CWWKO1752W: The OpenAPI UI Path has been set to the same path as the {0} feature and may prevent startup of itself or the other feature.
```
```suggestion
CWWKO1752W_OPEN_API_UI_PATH_POTENTIAL_CONFLICT=CWWKO1752W: Setting the OpenAPI UI Path to the same path as the {0} feature might disrupt the startup process of itself or the other feature.
```

helyarp replied:
Suggested change:
The OpenAPI UI Path was set to the same path as the {0} feature, which might prevent startup of itself or the other feature.

abutch3r replied:
suggested changes made

ramkumar-k-9286 commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.useraction=Specify a valid value for the extensions attribute in the configuration.
+
+CWWKO1752W_OPEN_API_UI_PATH_POTENTIAL_CONFLICT=CWWKO1752W: The OpenAPI UI Path has been set to the same path as the {0} feature and may prevent startup of itself or the other feature.
+CWWKO1752W_OPEN_API_UI_PATH_POTENTIAL_CONFLICT.explanation=The OpenAPI UI Path has been set to the same path as another feature and prevent startup or access.
```
```suggestion
CWWKO1752W_OPEN_API_UI_PATH_POTENTIAL_CONFLICT.explanation= If the OpenAPI UI Path shares the same path as another feature, it can prevent startup or access.
```

helyarp replied:
Suggested change:
The OpenAPI UI Path was set to the same path as another feature, which can prevent startup or access.

abutch3r replied:
Changes made as suggested

tevans78 commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    protected boolean validatePath(Config config, String path){
+        boolean valid=true;
+        //check against OpenAPI Data endpoint
+        String dataEndpoint = getOptionalValue(config, OPEN_API_ENDPOINT_CONFIG_NAME, String.class, OPEN_API_ENDPOINT_RESERVED_PATH);
+        if (path.equals(dataEndpoint)) {
+            Tr.warning(tc, "CWWKO1750W_OPEN_API_UI_PATH_RESERVED", OPEN_API_ENDPOINT_RESERVED_PATH);
+            valid=false;
+        }
+        //check against known invalid paths that contain valid characters
+        else if (Arrays.asList(INVALID_PATHS).contains(path)){
+            Tr.warning(tc, "CWWKO17510W_OPEN_API_UI_PATH_INVALID", path);
+            valid=false;
+        }
+        //check that path is made up of valid characters
+        else if(!PATH_PATTERN.matcher(path).matches()){
+            Tr.warning(tc, "CWWKO1751W_OPEN_API_UI_PATH_INVALID",path);
+            valid=false;
+        //check against other feature paths to see if a potential conflict might occur
+        }
+        return valid;
+    }
+
+    protected String resolvePath(String path){
+        // Add a forward slash if the path does not already start with one.
+        if (!path.startsWith("/")) {
+            path = '/' + path;
+        }
+        //in case someone specifies for example`////foo//bar` which will get through, keep running until all potential double `//` are accounted for
+        // given first pass will go to //foo/bar
+        while(path.contains("//")){
+            path = path.replace("//","/");
+        }
+        // Remove trailing slash if the path contains one.
+        if (path.endsWith("/") && !path.equals("/")) {
+            path = path.substring(0, path.length() - 1);
+        }
+        return path;
+    }
```
Should these three "processing" methods be static and have separate unit tests to check all the parsing edge cases and invalid path values?

abutch3r replied:
`validatePath` and `resolvePath` certainly should have unit tests. Though both were going to end up in common code when I did a bit of refactoring for the data endpoint as both have largely the same requirements
`validatePath` would need a small refactor to make it not reliant on config existing and that would allow it to also be `static`

The initial processing potentially could do, but would say moving the warning for potential conflicts out so that was more testable would be a better way to go.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+###############################################################################
+# Copyright (c) 2017 IBM Corporation and others.
```
Date

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED=CWWKO1750W: The OpenAPI UI Path cannot be set to {0}. This Path is reserved for use by the application server.
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.explanation=The OpenAPI UI Path cannot be set to a value that has been reserved for use by the application server.
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.useraction=Specify a valid value for the extensions attribute in the Micro Profile configuration.
```
This message is actually used when the OpenAPI document and UI paths are set to the same value but that isn't what it says.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/resources/com/ibm/ws/microprofile/openapi/ui/resources/OpenAPIUIMessages.nlsprops:
```diff
@@ -0,0 +1,34 @@
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.explanation=The OpenAPI UI Path cannot be set to a value that has been reserved for use by the application server.
+CWWKO1750W_OPEN_API_UI_PATH_RESERVED.useraction=Specify a valid value for the extensions attribute in the Micro Profile configuration.
+
+CWWKO1751W_OPEN_API_UI_PATH_INVALID=CWWKO1751W: The OpenAPI UI Path cannot be set to {0}. The Path is invalid or contains invalid characters.
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.explanation=The OpenAPI UI Path cannot be set to a value containing invalid characters.
+CWWKO1751W_OPEN_API_UI_PATH_INVALID.useraction=Specify a valid value for the extensions attribute in the configuration.
```
We should list the valid characters in the message, otherwise the user won't know what they've done that's not allowed.

posmith replied:
L2 reviewed and agrees that we should list the valid characters in the message.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+
+    private static final TraceComponent tc = Tr.register(OpenAPIUIEndpointManager.class);
+
+    private static final String OPEN_API_UI_VAR_Name = "openAPIUIPATH";
```
all caps for a constant

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+    protected void activate(ComponentContext context, Map<String, Object> properties){
+
+        String path = OPEN_API_UI_PATH;
+        if(ProductInfo.getBetaEdition()) {
```
We need an additional guard, since we're not even ready to beta this functionality yet.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+        openAPIWABConfigManager.deactivate();
+    }
+
+    protected String processOpenAPIUIPATH(final ComponentContext context, final Config config) {
```
"process" is not very descriptive of what this method does

Azquelt replied:
Why is this method `protected`?

Actually why are all these methods protected?

I think that `@Activate` and `@Deactivate` methods are generally supposed to be `protected` for some reason I can't remember, but I can't see why the others should be.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+
+        public void deactivateEvent() {
+            Tr.event(tc, "Unregistered UI web app bundle, WAB config=" + wabConfigReg + " for contextName=" + contextName + " and contextPath=" + contextPath);
+        }
```
I think the code would be clearer with these methods inlined, they're all only called from one place.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+        return optional.orElse(defaultValue);
+    }
+
+    final static class OpenAPIWABConfigManager{
```
I'm not sure what the point of having this as a separate inner class is.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+    public static <T> T getOptionalValue(Config config, String propertyName, Class<T> propertyType, T defaultValue) {
+        Optional<T> optional = config.getOptionalValue(propertyName, propertyType);
+        return optional.orElse(defaultValue);
+    }
```
I think the code would be clearer if this method was inlined.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UICustomPathTest.java:
```diff
@@ -0,0 +1,159 @@
+
+        ShrinkHelper.exportDropinAppToServer(server, war, DeployOptions.SERVER_ONLY);
+
+        ServerFileUtils serverFileUtils = new ServerFileUtils();
```
I wouldn't use this class, it doesn't seem widely used or maintained.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UICustomPathTest.java:
```diff
@@ -0,0 +1,159 @@
+
+        ServerFileUtils serverFileUtils = new ServerFileUtils();
+
+        String bootProps = serverFileUtils.getServerFileLoc(server) + "/bootstrap.properties";
+        // append to bootstrap.properties
+        FileWriter writer = new FileWriter(bootProps, true);
+        writer.append(PROPERTY_NAME + "=" + PATH_VALUE);
+        writer.append(System.getProperty("line.separator"));
+        writer.close();
+
+        String jvmOptions = serverFileUtils.getServerFileLoc(server)+"/jvm.options";
+        writer = new FileWriter(jvmOptions, true);
+        writer.append("-Dcom.ibm.ws.beta.edition" + "=" + "true");
+        writer.close();
```
Usually we would just write the bootstrap.properties file ourselves and put it in `/publish/servers/<servername>` if the content doesn't need to be dynamic.

Here it might make sense for it to be dynamic if we write more tests, so we could do it this way, but in that case we'll also have to be a little bit careful to avoid the config from one test contaminating another test.

If we do want to write a properties file directly, I would generally use `Properties.load()` and `.store()` and you should always use the try-with-resources syntax to close writers where possible.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UICustomPathTest.java:
```diff
@@ -0,0 +1,159 @@
+        assertNotNull("200 response line", testGet200Response);
+    }
+
+    /**
+     * Wait for an element to appear, using the default short timeout
+     *
+     * @param searchRoot the point to start the search
+     * @param by the locating mechanism
+     * @return the located element
+     * @throws TimeoutException if the element is not found within the timeout
+     */
+    private static WebElement waitForElement(SearchContext searchRoot, By by) {
+        return waitForElement(searchRoot, by, SHORT_WAIT);
+    }
+
+    /**
+     * Wait for an element to appear, using a custom timeout
+     *
+     * @param searchRoot the point to start the search
+     * @param by the locating mechanism
+     * @param timeout the timeout
+     * @return the located element
+     * @throws TimeoutException if the element is not found within the timeout
+     */
+    private static WebElement waitForElement(SearchContext searchRoot, By by, Duration timeout) {
+        return new FluentWait<>(searchRoot).ignoring(NoSuchElementException.class)
+                .withTimeout(timeout)
+                .until(c -> c.findElement(by));
+    }
```
These methods should move to a common class.

Azquelt commented on dev/io.openliberty.microprofile.openapi.ui.internal_fat/fat/src/io/openliberty/microprofile/openapi/ui/internal/fat/tests/UICustomPathTest.java:
```diff
@@ -0,0 +1,159 @@
+        driver = new RemoteWebDriver(chrome.getSeleniumAddress(), new ChromeOptions());
+    }
+
+    @Test
+    public void testCustomUIPath() {
+        driver.get("http://host.testcontainers.internal:" + server.getHttpDefaultPort() + PATH_VALUE);
+
+        // Check the title loads
+        WebElement title = waitForElement(driver, By.cssSelector("h2.title"), LONG_WAIT);
+        assertThat("Page title", title.getText(), Matchers.containsString("Generated API"));
+
+        // Check the headerbar colour
+        WebElement headerbar = waitForElement(driver, By.cssSelector("div.headerbar"));
+        assertEquals("Headerbar colour", "rgba(25, 28, 44, 1)", headerbar.getCssValue("background-color"));
+
+        // Check the headerbar has a background image. It's a data URL, so it's hard to assert it's actually correct, we just assert that there is one
+        WebElement headerbarWrapper = waitForElement(headerbar, By.cssSelector("div.headerbar-wrapper"));
+        assertThat("Headerbar image", headerbarWrapper.getCssValue("background-image"), startsWith("url(\"data:image/png"));
+
+        // Check we can see and open the operation
+        WebElement testGetOpBlock = waitForElement(driver, By.id("operations-default-get_test"));
+        WebElement testGetButton = testGetOpBlock.findElement(By.tagName("button"));
+        testGetButton.click();
+        WebElement testGet200Response = waitForElement(testGetOpBlock, By.cssSelector("tr.response[data-code=\"200\"]"));
+        assertNotNull("200 response line", testGet200Response);
+    }
```
We could probably make this code common as well if we're going to want to repeat it in multiple places.

After calling `driver.get`, the only thing that would need to be passed in is the `driver`.

Azquelt reviewed (changes requested):
I've left a few comments, we also need to format the code according to the formatting rules in the project.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,192 @@
+            valid=false;
+        }
+        //check against known invalid paths that contain valid characters
+        else if (Arrays.asList(INVALID_PATHS).contains(path)){
+            Tr.warning(tc, "CWWKO17510W_OPEN_API_UI_PATH_INVALID", path);
+            valid=false;
+        }
```
This check only looks for the path being `/..` or `/.`, we should probably reject the path if it has any segment which is `..` or `.`

posmith commented:
L2 message review completed

abutch3r commented:
#libby #build

abutch3r commented:
#libby

abutch3r commented:
#libby #build

abutch3r committed:
Update messages for missed updates

abutch3r commented:
#libby

helyarp reviewed (commented):
Looks good!

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,153 @@
+    private static final String OPEN_API_UI_PATH = "/ui";
+
+    private static final Pattern PATH_PATTERN = Pattern.compile("^(/[\\w./_-]*)?$");
+    private static final Map<String, String> CONFLICTING_PATHS_MAP = Stream.of(
+                                                                               new AbstractMap.SimpleEntry<>("/ibm/api", "OpenAPI"),
+                                                                               new AbstractMap.SimpleEntry<>("/health", "mpHealth"),
+                                                                               new AbstractMap.SimpleEntry<>("/metrics", "mpMetrics")).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
I'm still not convinced this is necessary. I'll open an issue to review it again before release.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,153 @@
+     */
+    public static boolean validatePath(String path, String reservedPath) {
+        boolean valid = true;
+        try {
+            //temporarily construct a URL using provided path to use the URL validation code.
+            new URL("https://localhost" + path);
+        } catch (MalformedURLException e) {
+            Tr.warning(tc, "CWWKO1754W_OPEN_API_UI_PATH_INVALID",e.getLocalizedMessage());
+            return false;
+        }
```
Not convinced this is needed. In addition, the message produced here is likely to be less useful than those produced by the following lines.

We'll address it under another issue.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,153 @@
+            Tr.warning(tc, "CWWKO1754W_OPEN_API_UI_PATH_INVALID",e.getLocalizedMessage());
+            return false;
+        }
+        if (path.equals(reservedPath)) {
+            Tr.warning(tc, "CWWKO1752W_OPEN_API_PATH_CONFLICT", path);
+            valid = false;
+        }
+        //check that path is made up of valid characters
+        if (!PATH_PATTERN.matcher(path).matches()) {
+            Tr.warning(tc, "CWWKO1754W_OPEN_API_UI_PATH_INVALID", path);
+            valid = false;
+            //check against other feature paths to see if a potential conflict might occur
+        }
+
+        // check that no segment is just `/.` or `/..`, `/.....` should be left as is as it has no special meaning
+        ArrayList<String> segments = new ArrayList<>(Arrays.asList(path.split("/")));
+        if (segments.contains(".") || segments.contains("..")) {
+            Tr.warning(tc, "CWWKO1755W_OPEN_API_UI_PATH_SEGMENT_INVALID");
+            valid = false;
+        }
```
These should be errors, not warnings. To be addressed in another issue.

Azquelt replied:
Need to update both the message codes and the call to `Tr.warning`.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,153 @@
+                    Tr.warning(tc,"CWWKO1751W_OPEN_API_PATH_UPDATE_FAILED");
+                } else if(!uiPath.equals(OPEN_API_ENDPOINT_PATH + OPEN_API_UI_PATH)){
+                    Tr.info(tc,"CWWKO1750I_OPEN_API_PATH_UPDATE",uiPath);
+                }
```
I don't like that `resolvePathFromConfig` is setting fields which this method then modifies.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIWABConfigManager.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * Active Web Bundle
+     */
+    public void activate() {
```
I don't think activate and deactivate are not good terms for what this class is doing. Those words already have a specific meaning for bundles and components in OSGi and this class is doing something different.

This class is setting a value for a WAB context path variable.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIWABConfigManager.java:
```diff
@@ -0,0 +1,95 @@
+
+public class OpenAPIWABConfigManager {
+
+    private static final TraceComponent tc = Tr.register(OpenAPIUIEndpointManager.class);
```
The class on this line needs updating.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIWABConfigManager.java:
```diff
@@ -0,0 +1,95 @@
+
+    private static final TraceComponent tc = Tr.register(OpenAPIUIEndpointManager.class);
+
+    final ComponentContext context;
+    final String contextName;
+    final String contextPath;
+    final String name;
+
+    ServiceRegistration<WABConfiguration> wabConfigReg;
```
I think these should all be private

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,153 @@
+
+        if (ProductInfo.getBetaEdition()) {
+            Config config = ConfigProvider.getConfig(OpenAPIUIEndpointManager.class.getClassLoader());
+            if (config.getOptionalValue("open_api_path_enabled", Boolean.class).orElse(false)) {
```
This should read from system properties, not MP Config

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIWABConfigManager.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Dictionary;
+import java.util.Hashtable;
+
+public class OpenAPIWABConfigManager {
```
Class needs javadoc to say what it does.

Possibly it could have a better name too.

abutch3r committed:
Updates based on feedback

Update System Property check to only pull from system properties

Update test to correctly set the system property

resolvePathfromConfig now does all the work in updating the data and UI paths

simplify name of WAB manager

Messages have not been updated as there is a new issue for that work

Azquelt commented:
Raised #25367 for the message issues

abutch3r commented:
#libby #build

Azquelt commented on dev/com.ibm.ws.microprofile.openapi.ui/src/com/ibm/ws/microprofile/openapi/ui/OpenAPIUIEndpointManager.java:
```diff
@@ -0,0 +1,154 @@
+            //check for system property `open_api_path_enabled` as additional guide - getBoolean returns `true` if the value exists and is set to `true`, if the value is `false`
+            if (Boolean.getBoolean("open_api_path_enabled")) {
+                resolvePathFromConfig(config);
+            }
```
Please move the `ConfigProvider.getConfig()` call inside the guard.

abutch3r committed:
move config retrieval inside of all guards

abutch3r commented:
No Docs changes since the last review

Azquelt commented:
Personal build does include the most recent changes.

abutch3r commented:
#libby

abutch3r commented:
While build only ran ~33% of test cases it did include the UI tests which did pass

abutch3r commented:
Being replaced by #25436 as this undoes 95% of the changes that were done here so to keep the commits cleaner this is being closed
